// build-app/dist/scripts/seed.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const pathHelper_1 = require("./../helpers/pathHelper");
const install_1 = require("../scripts/install");
const build_1 = require("../scripts/build");
const utils_1 = require("./../helpers/utils");
const envHelper_1 = require("../helpers/envHelper");
exports.default = {
    command: 'seed',
    describe: 'Initial project setup: installs dependencies and seeds data',
    handler: commandHandler,
    builder: commandBuilder
};
function commandBuilder(yargs) {
    return yargs.example('seed', 'Initial project setup: installs dependencies and seeds data.');
}
function commandHandler(argv) {
    return __awaiter(this, void 0, void 0, function* () {
        install_1.default.installAll();
        yield build_1.default.build('full');
        utils_1.default.runCommand('node', [pathHelper_1.default.serverRelative(envHelper_1.default.getTsBuildEntry()), 'run', 'seed'], {
            title: 'Seed Data',
            path: pathHelper_1.default.buildRelative('./'),
            showOutput: true
        });
    });
}
